### PR TITLE
Update requirements to make PAM packagable

### LINKED ISSUE #178: pytest fails when using python 3.11 (and updated dependencies)
I tried to use pam with python 3.11 - generating a population (mostly with the facility sampler) worked fine and was roughly 20% faster than using python 3.7 (wohoo!)

-> See https://github.com/markusstraub/pam/tree/py311, especially the `requirements.txt` with all current dependencies

However I noticed that a few unit tests fail. This is (I guess) due to the updated (geo)pandas-dependencies:

```
FAILED tests/test_00_utils.py::test_build_geodataframe_for_pt_person - AssertionError: Attributes of DataFrame.iloc[:, 11] (column name="seq") are different
FAILED tests/test_00_utils.py::test_build_geodataframe_for_cyclist - AssertionError: Attributes of DataFrame.iloc[:, 11] (column name="seq") are different
FAILED tests/test_00_utils.py::test_build_hhld_geodataframe - AssertionError: Attributes of DataFrame.iloc[:, 11] (column name="seq") are different
FAILED tests/test_00_utils.py::test_build_pop_geodataframe - AssertionError: Attributes of DataFrame.iloc[:, 11] (column name="seq") are different
FAILED tests/test_09_samplers.py::test_random_sample_point_from_multilinestring_random_seed - ValueError: Total of weights must be greater than zero
FAILED tests/test_09_samplers.py::test_random_sample_point_from_multipoint_random_seed - ValueError: Total of weights must be greater than zero
```

The fails in `test_00_utils.py` don't worry me at all, apparently pandas 1.3.x converted the `seq` parsed from the matsim xml to float, with pandas 1.5.x it is an int.

The fails in `test_09_samplers.py` are strange, but maybe these two tests need to be fixed anyways:
- `test_random_sample_point_from_multilinestring_random_seed` calls `sampler.sample_point_from_multipolygon`
- `test_random_sample_point_from_multipoint_random_seed` calls `sampler.sample_point_from_multipolygon`

@YannisZa git tells me you wrote (or last edited) these two tests, do you have an idea what could be wrong here?

### PR DESCRIPTION
**Pull request by brynpickering**
I have split up the requirements to ensure only those needed by the package can be installed as a basis. The others are now split into "docs" (for the notebooks) and "tests".

I've also added a conda environment file, although it is mostly duplication of the requirements text files and can be removed once s2sphere is available on the conda-forge channel (https://github.com/conda-forge/staged-recipes/pull/23041). 

Incidently fixed #178 to ensure tests pass. There were copy-paste errors in some of the tests.

### CODE REVIEW

File: tests/test_00_utils.py [hunk 1]
```diff
@@ -147,22 +150,25 @@ def correct_cyclist_geodataframe():
...
-    assert [[(round(p[0], 6), round(p[1], 6)) for p in ls.coords] for ls in gdf['geometry'].to_list()] == [
+    expected_coords = [
         [(0.123336, 51.537168), (0.175434, 51.567355)], [(0.175434, 51.567355), (0.183252, 51.574827)],
         [(0.183252, 51.574827), (0.250898, 51.559106)], [(0.250898, 51.559106), (0.250894, 51.559016)],
         [(0.250894, 51.559016), (0.080917, 51.539493)], [(0.080917, 51.539493), (0.080917, 51.539493)],
-        [(0.080917, 51.539493), (0.146117, 51.526088)], [(0.146117, 51.526088), (0.123336, 51.537168)]]
+        [(0.080917, 51.539493), (0.146117, 51.526088)], [(0.146117, 51.526088), (0.123336, 51.537168)]
+    ]
+    built_coords = [[(round(p[0], 6), round(p[1], 6)) for p in ls.coords] for ls in gdf['geometry'].to_list()]
+    np.testing.assert_allclose(built_coords, expected_coords, rtol=1e-3)
```
Comment (brynpickering) [inline tests/test_00_utils.py:165]: This is a reasonably lax tolerance to make up for the fact that the numbers all changed (potentially) non-negigibly. I'm not familiar enough with the data to know if the numbers have changed enough to suggest a problem. I think it's more likely just a change in the CRS conversion.

File: tests/test_00_utils.py [hunk 2]
```diff
@@ -147,22 +150,25 @@ def correct_cyclist_geodataframe():
...
 
 
 def test_build_geodataframe_for_cyclist(cyclist, correct_cyclist_geodataframe):
     gdf = cyclist.build_travel_geodataframe()
     gdf = gdf[correct_cyclist_geodataframe.columns]
-    assert_frame_equal(gdf, correct_cyclist_geodataframe)
+    assert_frame_equal(gdf, correct_cyclist_geodataframe, check_dtype=False)
```
Comment (brynpickering) [inline tests/test_00_utils.py:171]: This and others below allow a column to be either float and int and still be compared as being equal provided they are the same value (1 == 1.0). Other option would be to change the test geodataframe to have integer "seq" values, but this approach seems more maintainable.

File: tests/test_09_samplers.py [hunk 3]
```diff
@@ -1,10 +1,11 @@
-import pytest
 import random
-import pandas as pd
-import geopandas as gp
-from shapely.geometry import Polygon, Point, LinearRing, LineString, MultiLineString, MultiPolygon, MultiPoint
 from types import GeneratorType
 
+import geopandas as gp
```
Comment (brynpickering) [inline tests/test_09_samplers.py:4]: No change in packages, just [sorted](https://pycqa.github.io/isort/)

File: tests/test_09_samplers.py [hunk 4]
```diff
@@ -220,7 +221,7 @@ def test_random_sample_point_from_multilinestring_random_seed(fixed_seed):
     poly = MultiLineString([p1, p2])
     gdf = gp.GeoDataFrame(df, geometry=[poly]*3)
     sampler = spatial.RandomPointSampler(gdf.geometry, patience=0, seed=fixed_seed)
-    assert list(sampler.sample_point_from_multipolygon(gdf.geometry[0]).coords[0]) == [10.134364244112401, 10.847433736937234]
+    assert list(sampler.sample_point_from_multilinestring(gdf.geometry[0]).coords[0]) == [0.5374569764496049, 0.0]
```
Comment (brynpickering) [inline tests/test_09_samplers.py:224]: copy-paste error in the method called. Didn't raise an error previously as non-zero areas weren't _required_. An update in geopandas(?) changed that, at which point providing a `multilinestring` to a method expecting a `multipolygon` failed. After fixing the method, I just pasted in the values that comes as a result. Checking the multilinestring, this point is on it.

File: tests/test_09_samplers.py [hunk 5]
```diff
@@ -239,7 +240,7 @@ def test_random_sample_point_from_multipoint_random_seed(fixed_seed):
     poly = MultiPoint([p1, p2])
     gdf = gp.GeoDataFrame(df, geometry=[poly]*3)
     sampler = spatial.RandomPointSampler(gdf.geometry, patience=0, seed=fixed_seed)
-    assert list(sampler.sample_point_from_multipolygon(gdf.geometry[0]).coords[0]) == [10.0, 10.0]
+    assert list(sampler.sample_point_from_multipoint(gdf.geometry[0]).coords[0]) == [0.0, 0.0]
```
Comment (brynpickering) [inline tests/test_09_samplers.py:243]: After fixing the method, I just pasted in the values that comes as a result. Checking the multilinestring, this point is on it.

File: .github/workflows/build_pipeline.yml [hunk 6]
```diff
@@ -7,31 +7,17 @@ jobs:
...
-        export C_INCLUDE_PATH=/usr/include/gdal
-        sudo apt-get install ca-certificates
-        export CURL_CA_BUNDLE=/etc/ssl/certs/ca-certificates.crt
-        pip install GDAL==3.4.1
-        pip install -e .[planner]
+        micromamba-version: latest
+        environment-file: environment.yml
+        init-shell: bash
+        post-cleanup: all
+        cache-environment: true
```
Comment (brynpickering) [inline .github/workflows/build_pipeline.yml:16]: I'll admit, I don't know if this cache will pass between the two CI pipelines.

Comment (mfitz) [reply .github/workflows/build_pipeline.yml:20]: We have configured the repo such that CI builds need to be approved for first-time committers, so I guess we'll find out when I approve the build on this branch...

File: .github/workflows/build_pipeline.yml [hunk 7]
```diff
@@ -2,36 +2,28 @@ name: CI
...
+        create-args: flake8
+        post-cleanup: all
+        cache-environment: true
+      
+    - name: install PAM
+      run: pip install .
+
     - name: Lint with flake8
       run: |
-        pip install flake8
```
Comment (mfitz) [inline .github/workflows/build_pipeline.yml:34]: That reminds me about #144 ...

File: tests/test_00_utils.py (hunk 1 above)
Comment (mfitz) [reply tests/test_00_utils.py:165]: Is this the smallest tolerance that fixes the test?

I agree that this is likely a change in the CRS conversion across library versions. I'm not mad keen on the assertion in the first place, but that gripe predates this PR.

File: tests/test_00_utils.py (hunk 2 above)
Comment (mfitz) [reply tests/test_00_utils.py:171]: I definitely prefer this approach to modifying the dataframe types.

File: Dockerfile [hunk 8]
```diff
@@ -1,14 +1,9 @@
-FROM python:3.7.16-bullseye
+FROM mambaorg/micromamba:1.4.3
```
Comment (mfitz) [inline Dockerfile:1]: Glad to see the Dockerfile simplified. Hopefully, the resultant Docker image is smaller too - do you have the before and after stats?

File: Dockerfile [hunk 9]
```diff
@@ -1,14 +1,9 @@
...
-&& /usr/local/bin/python -m pip install --upgrade pip
+RUN pip install -e .
 
-COPY . .
-
-RUN pip install -e .[planner]
-ENV PYTHONPATH=./scripts:${PYTHONPATH}
-
-ENTRYPOINT ["python3"]
+ENTRYPOINT ["/usr/local/bin/_entrypoint.sh", "ipython"]
```
Comment (mfitz) [inline Dockerfile:9]: There are no instructions on using the Docker image in the README, and I don't get the impression people are using Dockerised PAM much, so this is not a question of great import, but...

Will this change how a user invokes anything via a container?

**Review by mfitz — APPROVED** [review]
Looks good.

We'll see how the CI builds go once it's approved and then merged...

File: tests/test_00_utils.py (hunk 1 above)
Comment (brynpickering) [reply tests/test_00_utils.py:165]: yeah, agreed, it's quite unmaintainable...
I think possibly `4e-4` or something like that is *the* smallest, but I like it being `1e...`

File: Dockerfile (hunk 8 above)
Comment (brynpickering) [reply Dockerfile:1]: Yup, it does make it smaller! locally, before=2.31GB, after=1.53GB

File: Dockerfile (hunk 9 above)
Comment (brynpickering) [reply Dockerfile:9]: Good question. I didn't have any examples to play with due to the lack of instructions in the README. I'll check with @Theodore-Chatziioannou. 

Anyway, the first element is just a requirement to have the mamba env variables persist (see [here](https://github.com/mamba-org/micromamba-docker#activating-a-conda-environment-for-entrypoint-commands)). I updated to `ipython` as it *should* provide the same and _more_ functionality as `python`.

**Comment by brynpickering** [comment]
I should note @mfitz that the CI builds on my fork are failing (e.g., [this one](https://github.com/brynpickering/pam/actions/runs/5256476241/jobs/9497857105)) due to issues with secrets and AWS access credentials (I assume both would be dealt with in the merge CI run as the secrets are set in this repo) AND coverage.

Locally, I have coverage >89%, but I think the `__init__.py` file I added isn't being correctly ignored by the run in the GH action. I have a PR that would clean up QA and for which the [tests _do_ pass](https://github.com/brynpickering/pam/actions/runs/5256611971/jobs/9498183085)

**Review by Theodore-Chatziioannou — APPROVED** [review]
managed to install PAM on Windows and WSL. My only deviation from the readme was using the "miniforge" prompt on Windows, and calling `miniconda` on WSL.

**Pull request merged**